### PR TITLE
[Solvers] Add active solvers page 

### LINKED ISSUE #221: [Solvers] Active solvers Table
As part of #156, we need to create  the Active solvers table. This will include: 

- Active solvers Table.
- Logic to fetch active solvers from the subgraph using the sdk.

Screen Shot 2022-09-20 at 22 35 28

Figma designs: https://www.figma.com/file/Y7pY6gGH2Wss6qhJwVcDJk/CoW-Explorer-(migrated)?node-id=0%3A1

### PR DESCRIPTION
**Pull request by matextrem**
# Summary

Closes #221 

Added Solvers page along with active solvers table.

Screen Shot 2022-09-22 at 23 32 07

# To Test

1.  Open the page `solvers`
      * You'll see the active solvers tab with the data in it.

### Notes

- Settlements tab is not implemented so pagination refers only to active solvers table.

### CODE REVIEW

**Comment by elena-zh** [comment]
Hey @matextrem , nice changes!
Some issues from my side: 
1. As far as I can understand, solvers list is displayed per network. So it would be great to show a network indicator in the URL when open the list in ethereum: 
image

2. it would be great to load solvers list when switch between networks. Currently, the app returns to the Home page when I switch a network
image

3. Sometimes it takes too long to load solvers list. And the query is running when I switch between tabs on the list (solvers/settlements). Is that OK? 
4. There is no default sort order on the table. It would be nice to add it by Total volume (and if no USD estimation, then Total settlements), and add an arrow to the column to show the sorting
5. I don't think we need 2 paginations on the table. 
2 paginations
6. It would be great to add a copy icon near the Solver address. WDYT?
add copy
7. I think, there is no need to show lots of decimals in USD tooltip. 2 is enough
image
8. The table is too wide in a mobile view. In addition, the menu is hidden
image
9. We tend to start the 2nd word from lower case letters across the app (User details/Tokens table. So it would be nice to apply it to the Tokens table as well
names from lowercase
10. there are 2 columns  (Name and Solver address). that navigate a user to address details on blockexplorer. Name column should navigate to a single solver details page (in the same tab)
11. There are different alignment in 2 columns on the page. I think, we should use the same alignment (left-aligned) for both. The same is for column names on reduced screens
diff alignment
image

Thanks!

**Comment by alongoni** [comment]
Hey @matextrem it's looking great! I would like to add a few comments:
1. Regarding point 5 from @elena-zh's feedback. I think that the two paginations are ok on a long page, so users don't need to navigate to the top to change the page. Also it would be great if we align all the paginators to the right.
https://www.loom.com/share/966688019ec34351907030b374118028
2. Related to point 8 from Elena's feedback. Tabs on mobile should look like this:
image
3. We need to add the Solver page link to the main menu. 
image

Thanks!!

**Comment by elena-zh** [comment]
@alongoni , 

> Regarding point 5 from @elena-zh's feedback. I think that the two paginations are ok on a long page, so users don't need to navigate to the top to change the page. Also it would be great if we align all the paginators to the right.
Agree, but then it would be great to add it to every table in the Explorer 😜

**Comment by matextrem** [comment]
Hey @elena-zh, regarding 1 and 2, we are doing the same as the rest of tables/pages. Ethereum has no prefix and changing networks redirects to the home page. In the case we wanted to change it, it should be done in a separate PR since it is not related to this one.

**Comment by elena-zh** [comment]
> Hey @elena-zh, regarding 1 and 2, we are doing the same as the rest of tables/pages. Ethereum has no prefix and changing networks redirects to the home page. In the case we wanted to change it, it should be done in a separate PR since it is not related to this one.

Hey @matextrem , agree, case 1 is irrelevant as an URL without a netowkr indicator always points to Ethereum data, and it is correct. 

In regard to the 2nd point, here is a video for you where User details **list** is still displayed when [change a network](https://watch.screencastify.com/v/RgbWT2LJZZ3tDFu1QZBH). 
I highlighted 'list', because you a currently working on the Solvers **list** page that exists in every network, not on a single order/transaction/solver/batch details that can't exist in another network (so that is why users are navigated to Home page from there on a network change). 
Thanks

**Comment by elena-zh** [comment]
@matextrem , in general, great job!

Some my notes what is left to fix: 
1.  default sort order on the table: it would be great to add 'Sorted by Total volume: from highest to lowest' message to mobile view (as we do for Top tokens table. 
2. It is a nitpick and can be skipped, but it would be great to add an additional sorting by Total settlements when Total volume = 0
image
3. I have just realized that there is no need to show a full USD amount on hover as it duplicates the original amount. WDYT?
image
3. Still, the 2nd word  starts with an upper-case letter in a desktop view. Also, it would be great to fix it on the tab name as well (both mobile and desktop views)
image
4. `there are 2 columns (Name and Solver address). that navigate a user to address details on blockexplorer. Name column should navigate to a single solver details page (in the same tab)` - not fixed
5. `We need to add the Solver page link to the main menu.`- not fixed. 

Thanks!

**Comment by alfetopito** [comment]
Great points Elena, agree with all of them.

Regarding 

> Hey @matextrem , agree, case 1 is irrelevant as an URL without a netowkr indicator always points to Ethereum data, and it is correct.

> In regard to the 2nd point, here is a video for you where User details list is still displayed when [change a network](https://watch.screencastify.com/v/RgbWT2LJZZ3tDFu1QZBH).
> I highlighted 'list', because you a currently working on the Solvers list page that exists in every network, not on a single order/transaction/solver/batch details that can't exist in another network (so that is why users are navigated to Home page from there on a network change).

Also agree. Some views make sense to redirect to home page when switching the network (order details, tx details etc) while some it makes sense to remain on the same page (user orders and solvers list)

But to keep the scope small, I'm fine to have that implemented as a different task/PR

File: src/hooks/useGetSolvers.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    }
+  }
+`
+
+export const GET_SETTLEMENTS_QUERY = gql`
+  query GetSettlements($solver: String) {
+    settlements(where: { solver: $solver }) {
+      id
+    }
+  }
```
Comment (ramirotw) [inline src/hooks/useGetSolvers.ts:103]: this query should be removed in favor of adding a `numberOfSettlements` counter field to the User entity as with this solution we will face the 1.000 array limit

File: src/components/solver/ActiveSolverTable/index.tsx [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+      <td>
+        <HeaderTitle>Trades</HeaderTitle>
+        <HeaderValue>{numberOfTrades}</HeaderValue>
+      </td>
+      <td>
+        <HeaderTitle>Total volume</HeaderTitle>
+        <HeaderValue>
+          <TextWithTooltip textInTooltip={`$${Number(solvedAmountUsd).toFixed(2) || 0}`}>
+            {Number(solvedAmountUsd)
+              ? formatPrice({ price: new BigNumber(solvedAmountUsd), decimals: 2, thousands: true })
```
Comment (ramirotw) [inline src/components/solver/ActiveSolverTable/index.tsx:207]: I believe it makes sense to use the `numberFormatter` for volume as we are doing on the homepage 
image

File: src/components/solver/ActiveSolverTable/index.tsx [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+        <HeaderValue>
+          <IdenticonWrapper>
+            <Identicon address={address} size="md" />
+            <BlockExplorerLink type="address" networkId={network} identifier={address} label={name} />
+          </IdenticonWrapper>
+        </HeaderValue>
+      </td>
+      <td>
+        <HeaderTitle>Trades</HeaderTitle>
+        <HeaderValue>{numberOfTrades}</HeaderValue>
```
Comment (ramirotw) [inline src/components/solver/ActiveSolverTable/index.tsx:200]: Could we have a formatter here? Same as `numberOfSettlements` down below, I think a thousands separator is important

File: src/apps/explorer/pages/Solver/data.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
+export const ACTIVE_SOLVERS = {
```
Comment (alfetopito) [inline src/apps/explorer/pages/Solver/data.ts:1]: I've created a way to fetch this info dynamically from Dune/Github https://github.com/cowprotocol/explorer/pull/226

File: src/hooks/useGetSolvers.ts (hunk 1 above)
Comment (matextrem) [reply src/hooks/useGetSolvers.ts:103]: Agree, but that is not implemented yet, right?

**Comment by elena-zh** [comment]
Hey @matextrem , great changes!
I have created a separate issue #230 to fix navigation when switch networks from Solvers page. 

Some remaining issues from my side: 
1.` there are 2 columns (Name and Solver address). that navigate a user to address details on blockexplorer. Name column should navigate to a single solver details page (in the same tab)` - not fixed
2. It would be great to start 'solvers' from the lower-case letter in the tab name
tab lowercase

Thanks!

File: src/hooks/useGetSolvers.ts (hunk 1 above)
Comment (ramirotw) [reply src/hooks/useGetSolvers.ts:108]: @GabrielCamba was working on this

**Comment by ramirotw** [comment]
@matextrem could you rebase this branch on top of develop to clean the commits?

**Comment by ramirotw** [comment]
@matextrem could you add the dollar sign to the tooltip? Also now that you have formatted the Trades value the tooltip with the full value makes sense

image

**Comment by matextrem** [comment]
> Hey @matextrem , great changes! I have created a separate issue #230 to fix navigation when switch networks from Solvers page.
> 
> Some remaining issues from my side: 1.` there are 2 columns (Name and Solver address). that navigate a user to address details on blockexplorer. Name column should navigate to a single solver details page (in the same tab)` - not fixed 2. It would be great to start 'solvers' from the lower-case letter in the tab name tab lowercase
> 
> Thanks!

As discussed in the standup, we'll have the solver name as it is now and update it once we have it the solver details page.

**Review by elena-zh — APPROVED** [review]
Hey @matextrem , then, changes LGTM!
Btw, I have added a note to the #91 issue to  implement a possibility to search for a solver by a value that exists in a name/address.

Also, a bit confused with the placeholder: can 'solver' be different from 'name'? If no, one of these words seems redundant and can be removed. 
image

Thanks!

**Comment by ramirotw** [comment]
@elena-zh 

1. Total settlements are capped to 100 as that's what is returned from the subgraph. Once this PR is merged and deployed https://github.com/cowprotocol/subgraph/pull/66#event-7487235097 we will have the correct value to display
2. Regarding the `Solver address` column, we may replace it with `Active since` or similar as we will have that information once https://github.com/cowprotocol/subgraph/pull/65#issuecomment-1260103739 is merged

I'm in favor of waiting for those deployments to be ready and then update this PR.

image

**Review by alfetopito — COMMENTED** [review]
A few comments on the code.

One general request: please merge/rebase against develop as there are lots of changes not related to this PR

**Comment by elena-zh** [comment]
Hey @matextrem , I have noticed the enhanced performance for the table when I switch between Solvers/Settlements tabs, thank you!

But still the table is loaded up to 10 seconds when I open it/reload the page. 
See the video: https://www.loom.com/share/8cc913366c1546fbbd9cadb0fb715e9d

Thanks

**Comment by elena-zh** [comment]
Hey @matextrem , I see that total settlements column shows nothing now
image

Is this OK?

**Comment by matextrem** [comment]
@elena-zh I've removed it so you can test the table is loaded faster. We are going to change the query to get the settlements in a more efficient way

File: src/hooks/useGetSolvers.ts (hunk 1 above)
Comment (ramirotw) [reply src/hooks/useGetSolvers.ts:101]: @matextrem please remove this query as well as the Settlement type if we are not using it anymore

**Review by ramirotw — APPROVED** [review]
approved with some caveats, the `Total settlements` column update and also changing the `Solver address` column for the `Active since` should be completed in a small follow-up PR once we have that data in the subgraph.

And please @matextrem change the base branch to a new `solvers` branch instead of develop so we put these changes, the Settlements page and also the Solvers details in a single branch until everything is ready.

**Comment by matextrem** [comment]
@ramirotw First query to the subgraph is taking too long to respond

Screen Shot 2022-10-06 at 09 59 24

**Review by alfetopito — APPROVED** [review]
Approved!
Only one minor issue regarding the chunk name

**Review by alongoni — APPROVED** [review]

**Pull request merged**